### PR TITLE
Collecting function symbols in an expression

### PR DESCRIPTION
**Pull request by ahumenberger**
Providing a function `function_symbols` for collecting all `FunctionSymbol` objects in an expression.

### CODE REVIEW

**Comment by certik** [comment]
@ahumenberger thanks for the PR.

@isuruf, can you please review it?

File: symengine/visitor.cpp [hunk 1]
```diff
@@ -134,6 +134,53 @@ set_basic free_symbols(const Basic &b)
...
+public:
+    set_basic s;
+    uset_basic v;
+
+    void bvisit(const Subs &x)
+    {
+        set_basic set_ = function_symbols(*x.get_arg());
+        for (const auto &p : x.get_variables()) {
+            set_.erase(p);
+        }
```
Comment (isuruf) [inline symengine/visitor.cpp]: FunctionSymbols are not in `x.get_variables()` so this whole method can be removed

File: symengine/visitor.cpp [hunk 2]
```diff
@@ -134,6 +134,53 @@ set_basic free_symbols(const Basic &b)
...
+                p->accept(*this);
+            }
+        }
+    }
+
+    void bvisit(const Basic &x)
+    {
+        if (is_a<FunctionSymbol>(x)) {
+            s.insert(x.rcp_from_this());
+        }
```
Comment (isuruf) [inline symengine/visitor.cpp]: Can you remove the above 3 lines and add a new method,
```
void bvisit(const FunctionSymbo &x) {
    s.insert(x.rcp_from_this());
}
```

Comment (ahumenberger) [reply symengine/visitor.cpp]: In order to visit also the arguments of the FunctionSymbol one needs the method below. But then the lower part of the method is copied from `void bvisit(const Basic &x)`. Which version do you prefer, the original one or the one below?
```    
    void bvisit(const FunctionSymbol &x) {
        s.insert(x.rcp_from_this());
        
        for (const auto &p : x.get_args()) {
            auto iter = v.insert(p->rcp_from_this());
            if (iter.second) {
                p->accept(*this);
            }
        }
    }
```

Comment (ahumenberger) [reply symengine/visitor.cpp]: @isuruf May I ask for an answer. I'm not sure which version you prefer.

**Comment by isuruf** [comment]
@ahumenberger, I've generalized your method to `atoms`. Let me know what you think?

**Comment by ahumenberger** [comment]
@isuruf, looks nice. Thanks!

**Pull request merged**